manuscript/13-Modules.md

```diff
@@ -456,13 +456,186 @@ This example loads `module.js` as a module instead of a script by passing a seco
+
+W> Node.js support for JavaScript modules hasn't been implemented at the time of my writing, so it's possible the details could change during implementation. Make sure to read the most current Node.js documentation about JavaScript module support.
+
+#### Module Resolution and Lookup Sequence
+
+The `.mjs` file extension signifies to Node.js that the file should be loaded as a JavaScript module. Node.js traditionally supported the `.js` and `.json` file extensions to load JavaScript and JSON files, respectively. The `.mjs` file extension has the highest precedence amongst the three file extensions, so Node.js will always look for `.mjs` files first whenever an extension is not present in the `import` statement (you must include the `.json` file extension explicitly to load a JSON file). Consider the following example:
+
+```js
+import { sum } from "./example";
+```
+
+When Node.js sees `"./example"`, a path without an extension, it searches the following locations in order, moving on to the next location if the previous one isn't found:
+
+1. `./example.mjs`
+1. `./example.js`
+1. `main` entry in `./example/package.json`
+1. `./example/index.mjs`
+1. `./example/index.js`
+
+This lookup sequence allows you to place JavaScript modules and CommonJS modules in the same directory, ensuring (but not requiring) backwards compatibility for other consumers. Those using an older version of Node.js would always load the `.js` file whereas a newer version would load the`.mjs` file. The lookup sequence is particular important in the case of packages installed in the `node_modules` directory, such as:
+
+```js
+import { sum } from "example";
+```
+
+In this case, `"example"` is a package name rather than a filename. As such, the lookup sequence is slightly different:
+
+1. `./node_modules/example.mjs`
+1. `./node_modules/example.js`
+1. `main` in `./node_modules/example/package.json`
+1. `./node_modules/example/index.mjs`
+1. `./node_modules/example/index.js`
+1. Repeat this process in the parent directory
+
+As with single files, packages can contain both `.mjs` and `.js` files to ensure compatibility with older versions of Node.js.
+
+A> #### Module Resolution Differences Between the Browser and Node.js
+A>
+A> The ECMAScript 6 standard does not specify the format of module filenames in `import` statements, leaving those details up to the implementations. As such, there's a subtle but important difference in the way browsers and Node.js resolve modules. Consider the following:
+A>
+A> ```js
+A> import { sum } from "example/sum.js";
+A> ```
+A>
+A> While this code may look simple to understand, browsers and Node.js interpret `"example/sum.js"` differently. The differences are outlined in the table below.
+A>
+A> | **Environment** | **File to Load** | **Loaded As** | **Mode** |
+A> -----------------------------------------------------------------
+A> | Browser         | `./example/sum.js` | JavaScript module | strict |
```

Comment:
Bare paths ala "example/" are banned by https://html.spec.whatwg.org/multipage/webappapis.html#resolve-a-module-specifier , but http://whatwg.github.io/loader/ will provide hooks to handle them.

So this does not load `./example/sum.js`, it errors out.